### PR TITLE
CSS: Ignore the CSS cascade in show()/hide()/etc.

### PR DESCRIPTION
**Pull request by gibson042**
Fixes gh-1767
Fixes gh-2071

No more showing or animating stylesheet-hidden elements. And please note the new use of `jQuery.swap` (which interacts with #2182).

```
   raw     gz Compared to master @ 062b5267d0a3538f1f6dee3df16da536b73061ea    
 -1335   -425 dist/jquery.js                                                   
  -433   -204 dist/jquery.min.js
```

### CODE REVIEW

**Comment by gibson042** [comment]
Ready for review. :fireworks:

File: src/css.js [hunk 1]
```diff
@@ -10,17 +10,16 @@ define([
 	"./css/var/getStyles",
 	"./css/curCSS",
 	"./css/adjustCSS",
-	"./css/defaultDisplay",
 	"./css/addGetHookIf",
 	"./css/support",
-	"./data/var/dataPriv",
+	"./css/showHide",
```
Comment (dmethvin) [inline src/css.js]: Can we pull `.show()` and `.hide()` into this file and remove the dependency inside css.js completely? That way it becomes a separate module and can be removed in a custom build if you use classes for showing and hiding.

Comment (gibson042) [inline src/css.js]: So, introduce a new showHide module (`.show`, `.hide`, and `.toggle`) depending on data and depended upon by effects? That sounds reasonable, but out of scope for this ticket.

Comment (dmethvin) [inline src/css.js]: Yeah, it can go in a different ticket and pull request, I'll create an issue. Now that `showHide` is broken out it seems pretty clear we could do that, and in fact the new module only helps with size reductions if it can be excluded. As it is right now, it can't ever be excluded because of the dependency here.

**Comment by dmethvin** [comment]
I love this PR, it simplifies things *so much*. The current code was really over-thinking so many edge cases and making life miserable.

It is a huge breaking change though, so it will need a lot of advance warning before 3.0 ships. All those edge cases are no doubt being used by people in various places, and we'll need to update the docs for these methods to clarify the cases we no longer support.

File: test/unit/css.js [hunk 2]
```diff
@@ -475,14 +475,14 @@ test("show(); hide()", function() {
 
 	hiddendiv = jQuery("div.hidden");
 	hiddendiv.hide();
-	equal( hiddendiv.css("display"), "none", "Non-detached div hidden" );
+	equal( hiddendiv.css("display"), "none", "Cascade-hidden div after hide()" );
 	hiddendiv.show();
-	equal( hiddendiv.css("display"), "block", "Pre-hidden div shown" );
+	equal( hiddendiv.css("display"), "none", "Show does not trump CSS cascade" );
```
Comment (timmywil) [inline test/unit/css.js]: This is the case that concerns me. I have used the stylesheet to hide elements initially and later called `.show()` on them. The new workflow – namely removing the `"hidden"` class – seems fine, but could require lots of changes by the user. We'll want to make sure this is covered in the blog post, docs, and migrate.

Comment (dmethvin) [inline test/unit/css.js]: I've added a note to the draft blog post on it. Adding this to Migrate will be fun since it has to add back all the crap above.

Comment (gibson042) [inline test/unit/css.js]: Yes, that is the crux of it, and the reason why #1767 is a blocker. It seems worth it to me, especially after seeing the code changes, but does require some substantial changes from everyone trying to show stylesheet-hidden elements.

**Comment by dmethvin** [comment]
This should reference #2057 and would resolve it as far as I'm concerned.

**Comment by timmywil** [comment]
@dmethvin Thanks Dave. I agree, and now it does. :)

**Comment by Mottie** [comment]
I would like to suggest adding a boolean (e.g. `.show(true)`) to mimic the current behavior and show cascade-hidden elements. A named option would also be required if the developer is using an object. Sadly, this idea would not work with `.toggle()`.

**Comment by gibson042** [comment]
I don't see that happening, for a few reasons:
* We're trying to avoid new boolean traps: https://github.com/jquery/jquery/wiki/Adding-new-features
* It's easy to handle with client code (e.g., `$el.removeClass("hidden").show()` or—at worst—`$el.css( "display", getVisibleDisplay ).show()`)
* We relied on [_**ugly**_ kludges](https://github.com/jquery/jquery/pull/2180/files#diff-aa200529dc721786f51c5f9e7af3d592) to determine substitute display values, [causing new problems](https://github.com/jquery/jquery/issues/2062) [without even guaranteeing correctness](http://bugs.jquery.com/search?q=defaultdisplay&go=&ticket=on)

**Comment by gibson042** [comment]
However, I just looked over the code again, and `defaultDisplay` may be separable from post-animation display restoration (which were the primary impetus for this PR). If it is, and if we're willing to pay the size and maintenance costs, it should be possible to bring back showing cascade-hidden elements without re-breaking gh-1767 and gh-2071 (except for initially-hidden elements, which are SOL regardless).

I'm still against it, but thought the idea should be on the record.
